### PR TITLE
#93: update to webpack@2 (closes #93)

### LINKED ISSUE #93: update to webpack@2
## requirements
as per title

## specs
- update webpack config & loaders

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by giogonzo**
Closes #93

## Notes
- This adds some time to the compilation :( see table here: #95
- About e2face8 , see https://webpack.js.org/configuration/node/ (quite scary.. not sure what's best to do)
- About letting webpack handle modules entirely: we'll need to change `babel-preset-buildo` or to configure it somehow via `scriptoni` (not really sure). For the time being, I just verified it works passing `"modules": false` via `.babelrc` on aliniq. These are real errors there (see https://github.omnilab.our.buildo.io/buildo/aliniq/issues/5596):
screen shot 2017-05-26 at 10 09 19 pm

### tests performed

Tested the following on these projects:
- [x] aliniq
- [x] gdsm
- [ ] bobafett
- [x] webseed

**dev-specific** (`npm start`)
- [x] after any (meaningful, and excluding 'config', html template, other config files...) change, the browser should refresh completely
- [x] an eslint error should be logged but doesn't fail the build
- [x] an eslint warning should be logged but doesn't fail the build
- [ ] first build time should not increase significantly ( 😢  see table here: https://github.com/buildo/scriptoni/pull/95)
- [ ] re-build time after a single change should not increase significantly ( 😢  see table here: https://github.com/buildo/scriptoni/pull/95)

**prod-specific** (`NODE_ENV=production npm run build`)
- [x] an eslint error should fail the build
- [x] an eslint warning should fail the build
- [x] bundle size is not (significantly) different
- [x] gzipped bundle size is not (significantly) different (`du -h build/bundle.<hash>.js`)

**either dev or prod**
- [x] source maps are working

**bundle before/after**
*(also post bundle before if there are any relevant changes)*
*this can be easily obtained with `source-map-explorer build/bundle.<hash>.js`*
screen shot 2017-05-26 at 8 38 07 pm

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.}

### CODE REVIEW

File: src/scripts/webpack/webpack.base.js [hunk 1]
```diff
@@ -10,24 +10,30 @@ const JSLoader = t.enums.of(['babel', 'typescript'], 'JSLoader');
...
+        options: {
+          failOnError: NODE_ENV === 'production',
+          failOnWarning: NODE_ENV === 'production'
+        }
+      }],
       include: paths.SRC,
       exclude: paths.ASSETS
     }
-  ] : [];
+  ];
```
Comment (gabro) [inline src/scripts/webpack/webpack.base.js]: I would just move the "preloaders" down with the rest of the rules, now that they're not optionally included

Comment (giogonzo) [inline src/scripts/webpack/webpack.base.js]: sure 👍

**Comment by gabro** [comment]
Cool! Did you solve the issues with sass variable loader and table row heights too?

**Comment by giogonzo** [comment]
@gabro no it turns out it was not related to sass-variables, it was just: https://github.com/buildo/scriptoni/commit/e2face869336a3dd87404448a4f3dd48bd8cf127 (see the first **note** above)

**Comment by giogonzo** [comment]
(we were importing as `constants` a node built-in module instead of our own `constants.js` file :/)

**Comment by gabro** [comment]
Ah, damn.
I thought require('./constants') would work thought, were we doing require('constants)?

--
gabriele
~ buildo

**Comment by giogonzo** [comment]
since it's in `src/app/constants.js` and we have `src/app` as webpack `resolve` root, you can include it as just `constants` from everywhere. These used to have precedence over anything else, but something must be changed with newer webpacks..

**Comment by gabro** [comment]
From the discussion I've read, node native modules always had the precedence. Webpack 2 simply added constants as a native module, shadowing all other imports

--
gabriele
~ buildo

**Comment by gabro** [comment]
Cfr https://github.com/webpack/webpack/issues/4666#issuecomment-292700060

**Comment by gabro** [comment]
what other projects use `scriptoni webpack` for building?

**Comment by veej** [comment]
> what other projects use scriptoni webpack for building?

I'll test asap on bobafett... should be the last one

**Comment by gabro** [comment]
how about lexdoit?

**Comment by giogonzo** [comment]
@gabro not using scriptoni webpack (because SSR)

**Comment by giogonzo** [comment]
@veej 

> Build time:

are your measures including uglify+gzip or not? NODE_ENV?

**Comment by veej** [comment]
> are your measures including uglify+gzip or not? NODE_ENV?

whoops, tested with `NODE_ENV=development`

**Comment by giogonzo** [comment]
@veej no worries, just curious

OK then I'll merge this and prepare PRs with updated version on each project

**Pull request merged**